[GTK][WPE][OpenXR] Support running OpenXR API tests on bots

spenap opened the pull request:
#### 29adb1630b2ab13438013d0a2fcd64a7d2bd6fc8
<pre>
[GTK][WPE][OpenXR] Support running OpenXR API tests on bots
<a href="https://bugs.webkit.org/show_bug.cgi?id=297261">https://bugs.webkit.org/show_bug.cgi?id=297261</a>

Reviewed by Carlos Alberto Lopez Perez.

This adds support for running OpenXR API tests to the bots using
monado-service.

A new Monado driver (in a ~loose way) is added. This checks if
monado-service is available on the given port, and launches it with
XRT_COMPOSITOR_NULL, removing any dependency against Wayland, X11 or
other compositors, and so allowing run-{gtk,wpe}-tests to still handle
the --display-server they receive.

The existing TestWebKitImmersiveMode.cpp is renamed to
TestWebKitWebXR.cpp, allowing the test runner to check is_openxr_test
by name. The TestExpectations are updated to mark the test as passing on
WPE, where the bots have been updated to support monado-service, the GTK
ones will follow later.

* Tools/Scripts/webkitpy/port/monadodriver.py: Added.
(MonadoDriver):
(MonadoDriver.check_driver):
(MonadoDriver.__init__):
(MonadoDriver._setup_environ_for_test):
(MonadoDriver.stop):
* Tools/Scripts/webkitpy/port/monadodriver_unittest.py: Added.
(MonadoDriverTest):
(MonadoDriverTest.make_driver):
(MonadoDriverTest.test_start):
(MonadoDriverTest.test_stop):
(MonadoDriverTest.test_stop.FakeMonadoServiceProcess):
(MonadoDriverTest.test_stop.FakeMonadoServiceProcess.terminate):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebXR.cpp: Renamed from Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitImmersiveMode.cpp.
(WebXRTest::isImmersiveModeEnabledChanged):
(WebXRTest::permissionRequestCallback):
(testWebKitWebXRLeaveImmersiveModeAndWaitUntilImmersiveModeChanged):
(beforeAll):
* Tools/TestWebKitAPI/glib/CMakeLists.txt:
* Tools/TestWebKitAPI/glib/TestExpectations.json:
* Tools/glib/api_test_runner.py:
(TestRunner.__init__):
(TestRunner._tear_down_testing_environment):
(TestRunner._run_test_glib):
(TestRunner.is_webxr_test):
(TestRunner.run_tests):

Canonical link: <a href="https://commits.webkit.org/300454@main">https://commits.webkit.org/300454@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2fe0d1bf0418a2c1f0becf6e02abb04eab1b57a4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121756 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41457 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32121 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/128289 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73846 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/123632 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42169 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50048 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92517 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61507 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124708 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33643 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109044 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73181 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/121113 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32655 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27208 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71803 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103143 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27393 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131071 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48691 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37031 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101107 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49063 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105262 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100980 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25793 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46343 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24457 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/45392 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48549 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/48019 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51368 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49701 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

clopez commented on Tools/Scripts/run-gtk-tests:
```diff
@@ -91,6 +91,9 @@ class GtkTestRunner(TestRunner):
 
+    def is_openxr_test(self, test_program):
+        return "Immersive" in os.path.basename(test_program)
+
```
Since this function is the same fot GTK and WPE it can be moved to the base class `TestRunner` in Tools/glib/api_test_runner.py

clopez commented on Tools/Scripts/run-wpe-tests:
```diff
@@ -51,6 +51,9 @@ class WPETestRunner(TestRunner):
 
+    def is_openxr_test(self, test_program):
+        return "Immersive" in os.path.basename(test_program)
+
```
ditto

clopez commented on Tools/Scripts/webkitpy/port/monadodriver.py:
```diff
@@ -0,0 +1,65 @@
+
+import logging
+import os
+
+from webkitpy.port.driver import Driver
+
+_log = logging.getLogger(__name__)
+
+
+class MonadoDriver(Driver):
+    @staticmethod
+    def check_driver(port):
+        monado_findcmd = ['which', 'monado-service']
+        monado_found = port.host.executive.run_command(monado_findcmd, return_exit_code=True) == 0
+        if not monado_found:
+            _log.error("No monado-service found. Cannot run OpenXR API tests.")
+        return monado_found
+
+    def __init__(self, *args, **kwargs):
+        Driver.__init__(self, *args, **kwargs)
+
+    def _setup_environ_for_test(self):
+        driver_environment = super(MonadoDriver, self)._setup_environ_for_test()
+        driver_environment['WITH_OPENXR_RUNTIME'] = 'y'
+
+        monado_command = ['monado-service']
+        with open(os.devnull, 'w') as devnull:
+            self._monado_service_process = self._port.host.executive.popen(monado_command, stderr=devnull, env=driver_environment)
+
+        return driver_environment
+
+    def stop(self):
+        super(MonadoDriver, self).stop()
+        if getattr(self, '_monado_service_process', None):
+            self._monado_service_process.terminate()
+            self._monado_service_process = None
```
I'm not familiar with Monado, but perhaps `webkitpy.port.driver.Driver` is not the best place for implementing this.

Let me expain:

`webkitpy.port.driver.Driver` is meant to implement display-server drivers like:
```
  "xvfb": Use a virtualized X11 server.
  "xorg": Use the current X11 session.
  "weston": Use a virtualized Weston server (weston over xvfb).
  "wayland": Use the current wayland session.
  "headless": Use WPE headless.
```

So the idea of this drivers is to take care of starting and stopping the facilities to run WebKit inside a specific display-server: headless, or inside wayland, or inside xvfb, etc.

But this does not support stacking. You can select only one display-server driver. So that means that if you now implement a driver named Monado and you force to run the tests inside that `Monado` display-server that means no other display-server can be selected by the user. 

See for example the output of `Tools/Scripts/run-gtk-tests --help`: the user is able to select running the API tests inside their own X11 or Wayland session so they are able to see the test window draw on their session. Same with wpe: they can select to run headless or inside their wayland session.

Many of the API tests don't drawn anything at all, so this doesn't have effect.

If Monado tests don't draw anything at all the future expectation is the same (do not draw anything at all) then this approach can work, but if you want the possibility of allowing the user to select executing Monado tests headless or non-headless (see the window) then a different approach than using webkitpy.port.driver.Driver should be used.

spenap replied:
Thanks for this explanation. I am still using a driver as the base class for the Monado logic -- simply because it fitted so well with the start, stop and environment behaviour it needs. 

I have updated the logic so that running OpenXR tests no longer requires Weston -- nor any other display server, so the user can now pass whatever they want into `--display-server` without any interference, and without any visual output: Monado will use a _null compositor_.

clopez commented on Tools/glib/api_test_runner.py:
```diff
@@ -31,6 +31,7 @@
 from webkitpy.common.host import Host
 from webkitpy.common.test_expectations import TestExpectations
+from webkitpy.port.monadodriver import MonadoDriver  # noqa: E402
 from webkitpy.port.westondriver import WestonDriver
```
Here the Wayland tests are relying on WestonDriver (which executes Weston inside Xvfb) to force the Wayland tests to always run headless. Ideally the runner should only force the WestonDriver by default, but not if the user opted to select the Wayland runner

spenap replied:
I have updated the PR, so it no longer requires any display driver

clopez reviewed (commented):
Patch looks good overall.

clopez reviewed (approved):
LGTM. Thanks!

spenap committed:
[GTK][WPE][OpenXR] Support running OpenXR API tests on bots
https://bugs.webkit.org/show_bug.cgi?id=297261

Reviewed by Carlos Alberto Lopez Perez.

This adds support for running OpenXR API tests to the bots using
monado-service.

A new Monado driver (in a ~loose way) is added. This checks if
monado-service is available on the given port, and launches it with
XRT_COMPOSITOR_NULL, removing any dependency against Wayland, X11 or
other compositors, and so allowing run-{gtk,wpe}-tests to still handle
the --display-server they receive.

The existing TestWebKitImmersiveMode.cpp is renamed to
TestWebKitWebXR.cpp, allowing the test runner to check is_openxr_test
by name. The TestExpectations are updated to mark the test as passing on
WPE, where the bots have been updated to support monado-service, the GTK
ones will follow later.

* Tools/Scripts/webkitpy/port/monadodriver.py: Added.
(MonadoDriver):
(MonadoDriver.check_driver):
(MonadoDriver.__init__):
(MonadoDriver._setup_environ_for_test):
(MonadoDriver.stop):
* Tools/Scripts/webkitpy/port/monadodriver_unittest.py: Added.
(MonadoDriverTest):
(MonadoDriverTest.make_driver):
(MonadoDriverTest.test_start):
(MonadoDriverTest.test_stop):
(MonadoDriverTest.test_stop.FakeMonadoServiceProcess):
(MonadoDriverTest.test_stop.FakeMonadoServiceProcess.terminate):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebXR.cpp: Renamed from Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitImmersiveMode.cpp.
(WebXRTest::isImmersiveModeEnabledChanged):
(WebXRTest::permissionRequestCallback):
(testWebKitWebXRLeaveImmersiveModeAndWaitUntilImmersiveModeChanged):
(beforeAll):
* Tools/TestWebKitAPI/glib/CMakeLists.txt:
* Tools/TestWebKitAPI/glib/TestExpectations.json:
* Tools/glib/api_test_runner.py:
(TestRunner.__init__):
(TestRunner._tear_down_testing_environment):
(TestRunner._run_test_glib):
(TestRunner.is_webxr_test):
(TestRunner.run_tests):

Canonical link: https://commits.webkit.org/300454@main
